## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [28]:
import pandas as pd 

df = pd.read_csv("client_dataset.csv")

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [29]:
# Use this space to do any additional research and familiarize yourself with the data.

In [30]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [31]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [32]:
#df.rows
rows =len(df)
rows

54639

In [33]:
#df.dtypes
df.dtypes

first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [35]:
df.head(2)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   

   order_week  order_year         item_id     category subcategory  \
0          17        2023  EUD29711-63-6U        decor    wall art   
1          20        2023  XDA18116-89-4A  consumables        pens   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0

In [36]:
df.tail(2)

first     last                      job              phone  \
54637   Wanda  Solomon             Toxicologist      (311)767-4924   
54638  Dwayne     Bell  Chief Financial Officer  606-619-7667x9508   

                           email  client_id  order_id  order_date  order_week  \
54637  solomonwanda5962@ross.org      24485   7592730  2023-01-18           3   
54638           belld@galvan.com      24007   7489403  2023-01-04           1   

       order_year         item_id   category         subcategory  unit_price  \
54637        2023  XMZ38239-49-3D  furniture              stands        1.48   
54638        2023  ZCZ05589-50-5D   software  project management        3.01   

       unit_cost  unit_weight   qty  line_number  
54637       1.22        18.04   112            0  
54638       1.61         2.07  1031            5

In [37]:
#Correctly identify the category with the most entries. (4 points)
max_category = df['category'].value_counts().idxmax()
max_category

'consumables'

In [38]:
# What three item categories had the most entries?

df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [39]:
# For the category with the most entries,
# which subcategory had the most entries?

subcategory_df =  df.loc[(df['category'] == 'consumables')]
subcategory_df['subcategory'].value_counts().head(1)
### subcat_df = df.loc(consumables).... continue with cubcat_df[value...]


subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [40]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [41]:
# Store the client ids of those top 5 clients in a list.
list(df['client_id'].value_counts().head(5).index)


[33615, 66037, 46820, 38378, 24741]

In [42]:
# How many total units (the qty column) did the, client with the most entries order order?
#Display the total units (the qty column) that the client with the most entries ordered. (4 points)
totalunits_df = df.loc[df['client_id'] == 33615, 'qty'].sum()
totalunits_df


64313

In [43]:
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [44]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
#Create a column that calculates the subtotal for each line using the unit_price and the qty. (6 points)
df['line_subtotal'] = df['unit_price'] * df['qty']
df[['unit_price', 'qty', 'line_subtotal']].head()

unit_price  qty  line_subtotal
0     1096.80  105      115164.00
1       24.95   21         523.95
2       13.52   39         527.28
3       36.42   29        1056.18
4      195.10   20        3902.00

In [45]:
# Create a column for total weight.
df['total_weight'] = df['unit_weight'] * df['qty']

# Create a column for order weight.
df['t_order_weight'] = df['total_weight']

# Define a function to calculate shipping rate based on order weight.
def shipping_rate(t_order_weight):
    if t_order_weight > 50:
        return t_order_weight * 7
    else:
        return t_order_weight * 10

# Apply the shipping_rate function to the 'order_weight' column and create a new column 'final_shipping'.
df['shipping_rate'] = df['t_order_weight'].apply(shipping_rate)

# Create a column for shipping rate (corrected).
df['final_shipping'] = df['t_order_weight'].apply(shipping_rate)

# Display selected columns for the first 3 rows.
selected_columns = ['unit_price', 'unit_weight', 'qty', 'total_weight', 'final_shipping', 'shipping_rate']
df[selected_columns].head(3)

unit_price  unit_weight  qty  total_weight  final_shipping  shipping_rate
0     1096.80         7.50  105        787.50         5512.50        5512.50
1       24.95         1.49   21         31.29          312.90         312.90
2       13.52         1.68   39         65.52          458.64         458.64

In [46]:
# along with a sales tax of 9.25%
# representing 9.25% as a decimal
sales_tax_rate = 0.0925

# using the subtotal and the shipping price
df['total_price'] = (df['line_subtotal'] + df['shipping_rate']) * (1 + sales_tax_rate)

# Round the 'total_price' column to two decimal places
df['total_price'] = df['total_price'].round(2)

# Display selected columns for the first 3 rows
df[['line_subtotal', 'shipping_rate', 'total_price']].head(3)


line_subtotal  shipping_rate  total_price
0      115164.00        5512.50    131839.08
1         523.95         312.90       914.26
2         527.28         458.64      1077.12

In [47]:
# Create a column for the cost of each line using unit cost, qty, and shipping rate
df['line_cost'] = df['unit_cost'] * df['qty'] + df['shipping_rate']

# Display selected columns for the first 3 rows
selected_columns = ['line_cost', 'unit_cost', 'qty', 'shipping_rate', 'total_price']
df[selected_columns].head(3)

line_cost  unit_cost  qty  shipping_rate  total_price
0   85597.05     762.71  105        5512.50    131839.08
1     629.79      15.09   21         312.90       914.26
2     765.18       7.86   39         458.64      1077.12

In [48]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['total_price'] - df['line_cost']
df[['line_profit', 'total_price', 'line_cost']].head(3)


line_profit  total_price  line_cost
0     46242.03    131839.08   85597.05
1       284.47       914.26     629.79
2       311.94      1077.12     765.18

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [49]:
# Check your work using the totals above

def find_total_price(order_id):
    return df.loc[df['order_id'] == order_id, 'total_price'].sum()

order_ids_to_check = [2742071, 2173913, 6128929]

for order_id in order_ids_to_check:
    total_price_found = find_total_price(order_id)
    print(f"Order {order_id} total: ${total_price_found:.2f}")

Order 2742071 total: $152811.89
Order 2173913 total: $162388.71
Order 6128929 total: $923441.25


In [50]:
### NOTE: THE BELOW LISTED CODE: 
#df['total_price'] = (df['line_subtotal'] + df['shipping_rate']) * (1 + sales_tax_rate)

#RETURNS A VALUE: $152,811.89. while this is correct according to the assignment.

#This is impractical. In all 50 states, shipping fees are a service charge and not a purchase, therby they are exempt from sales taxes.
#to account for this the code should be written as below:

#df['total_price'] = (df['line_subtotal']* (1 + sales_tax_rate)) + df['shipping_rate']

#this returns the following values:
# Order 2742071 total: $151695.02; Order 2173913 total: $158762.83; Order 6128929 total: $908947.31'''

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [51]:
# How much did each of the top 5 clients, by quantity, spend?
# Check your work from Part 1 for client ids.
top5_client_ids = df.groupby('client_id')['total_price'].sum().nlargest(5).index

def find_total_price(client_id):
    return df.loc[df['client_id'] == client_id, 'total_price'].sum()

for client_id in top5_client_ids:
    total_spent = find_total_price(client_id)
    print(f'{client_id}: ${total_spent:,.2f}')


78965: $655,923,670.05
85931: $237,417,001.41
38473: $183,552,571.81
86721: $181,266,297.05
84344: $106,648,111.60


In [52]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. Sort by total profit.
client_summary = df.groupby('client_id').agg({'qty': 'sum', 'shipping_rate': 'sum', 'total_price': 'sum', 'unit_cost': 'sum'})

client_summary['total_profit'] = client_summary['total_price'] - client_summary['unit_cost']

client_summary = client_summary.sort_values(by='total_profit', ascending=False)

client_summary['total_profit'] = client_summary['total_profit'].map('${:,.2f}'.format)

top5_client_summary = client_summary.head(5)

print(top5_client_summary)


               qty  shipping_rate   total_price  unit_cost     total_profit
client_id                                                                  
78965      4033302     5340756.88  6.559237e+08   16420.61  $655,907,249.44
85931       644423    10122549.19  2.374170e+08    5592.40  $237,411,409.01
38473       402770     4391003.73  1.835526e+08   15872.48  $183,536,699.33
86721      1010426    64335727.72  1.812663e+08   12976.28  $181,253,320.77
84344       466968     4746583.13  1.066481e+08   13943.23  $106,634,168.37


In [53]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. Sort by total profit.
client_summary = df.groupby('client_id').agg({'qty': 'sum', 'shipping_rate': 'sum','total_price': 'sum', 'unit_cost': 'sum'})

# Sort by total profit.
client_summary['total_profit'] = client_summary['total_price'] - client_summary['unit_cost']
client_summary['total_profit'] = client_summary['total_profit'].map('${:,.2f}'.format)
client_summary['unit_cost'] = client_summary['unit_cost'].map('${:,.2f}'.format)
client_summary['total_price'] = client_summary['total_price'].map('${:,.2f}'.format)
client_summary['shipping_rate'] = client_summary['shipping_rate'].map('${:,.2f}'.format)

client_summary = client_summary.sort_values(by='total_profit', ascending=False)
top5_client_summary = client_summary.head(5)
print(top5_client_summary)

            qty shipping_rate    total_price  unit_cost total_profit
client_id                                                           
24502      7794   $207,042.26  $1,003,911.11  $4,637.57  $999,273.54
31885      2759    $67,887.90  $1,002,141.86  $2,899.02  $999,242.84
20820      4229   $227,584.88  $1,001,848.35  $3,425.30  $998,423.05
33475      5814   $196,320.77    $994,348.06  $2,613.41  $991,734.65
17425       897    $17,799.71    $100,176.93    $727.71   $99,449.22


In [54]:
# Format the data and rename the columns to names suitable for presentation.

client_summary = client_summary.rename(columns={'qty': 'Total Units Purchased', 'shipping_rate': 'Total Shipping Cost',
                                                'total_price': 'Total Revenue', 'unit_cost': 'Total Cost', 
                                                'total_profit': 'Total Profit'})

top5_client_summary = client_summary.head(5)
top5_client_summary.head(5)


Total Units Purchased Total Shipping Cost  Total Revenue  \
client_id                                                             
24502                       7794         $207,042.26  $1,003,911.11   
31885                       2759          $67,887.90  $1,002,141.86   
20820                       4229         $227,584.88  $1,001,848.35   
33475                       5814         $196,320.77    $994,348.06   
17425                        897          $17,799.71    $100,176.93   

          Total Cost Total Profit  
client_id                          
24502      $4,637.57  $999,273.54  
31885      $2,899.02  $999,242.84  
20820      $3,425.30  $998,423.05  
33475      $2,613.41  $991,734.65  
17425        $727.71   $99,449.22

In [55]:
# write a brief 2-3 sentence summary of your findings
"""after reviewing the data and following all steps, I found the data results lacking some pertinent data. While it gave some decent results,
it lacked demographics, beyond a client ID number, cashflow and profit. This is good information for checking company status and some financial reporting.

It doesn't say who those clients were; private party or Business-to-business; what they do; what industry they are in.
Therefore the data reported is useless for production & logistics planning. 

I've written some code below to add some basic & aggregated data  to put a face, market/industry and production/procurement/logistical planning based 
off of the top5 client IDs. Combined, both datasets, will enable the company to look into areas improvement, expansion and development."""

top5_client_demo = pd.merge(client_summary.head(5), df[['client_id', 'first', 'last', 'job', 'category', 'total_price', 'line_profit']], on='client_id')
top5_client_demo_aggregated = top5_client_demo.groupby(['client_id', 'first', 'last', 'job']).agg({
    'total_price': 'sum',
    'line_profit': 'sum'
}).reset_index()
top5_client_demo_aggregated['total_price'] = top5_client_demo_aggregated['total_price'].map('${:,.2f}'.format)
top5_client_demo_aggregated['line_profit'] = top5_client_demo_aggregated['line_profit'].map('${:,.2f}'.format)
top5_client_demo_aggregated = top5_client_demo_aggregated.sort_values(by='line_profit', ascending=False)
top5_client_demo_aggregated[['client_id', 'first', 'last', 'job', 'total_price', 'line_profit']]


client_id      first     last                      job    total_price  \
4      33475  Christine   Crosby         Technical brewer    $994,348.06   
3      31885    Melissa   Harris           Engineer, site  $1,002,141.86   
2      24502     Daniel   Deleon  Rural practice surveyor  $1,003,911.11   
1      20820      Laura  Nichols               Air broker  $1,001,848.35   
0      17425     Sylvia   Duncan  Secretary/administrator    $100,176.93   

   line_profit  
4  $348,654.15  
3  $328,118.40  
2  $312,834.19  
1  $226,515.85  
0   $19,141.27